yjit/src/codegen.rs

```diff
@@ -6961,114 +6958,189 @@ fn gen_send_iseq(
         // This is the list of keyword arguments that the callee specified
         // in its initial declaration.
         let callee_kwargs = unsafe { (*keyword).table };
-        let total_kwargs: usize = unsafe { (*keyword).num }.try_into().unwrap();
+        let callee_kw_count_i32: i32 = unsafe { (*keyword).num };
+        let callee_kw_count: usize = callee_kw_count_i32.try_into().unwrap();
+        let keyword_required_num: usize = unsafe { (*keyword).required_num }.try_into().unwrap();
 
         // Here we're going to build up a list of the IDs that correspond to
         // the caller-specified keyword arguments. If they're not in the
         // same order as the order specified in the callee declaration, then
         // we're going to need to generate some code to swap values around
         // on the stack.
-        let mut caller_kwargs: Vec<ID> = vec![0; total_kwargs];
-
+        let mut kwargs_order: Vec<ID> = vec![0; cmp::max(caller_keyword_len, callee_kw_count)];
         for kwarg_idx in 0..caller_keyword_len {
             let sym = unsafe { get_cikw_keywords_idx(ci_kwarg, kwarg_idx.try_into().unwrap()) };
-            caller_kwargs[kwarg_idx] = unsafe { rb_sym2id(sym) };
+            kwargs_order[kwarg_idx] = unsafe { rb_sym2id(sym) };
         }
-        let mut kwarg_idx = caller_keyword_len;
 
         let mut unspecified_bits = 0;
 
-        let keyword_required_num: usize = unsafe { (*keyword).required_num }.try_into().unwrap();
-        for callee_idx in keyword_required_num..total_kwargs {
-            let mut already_passed = false;
-            let callee_kwarg = unsafe { *(callee_kwargs.offset(callee_idx.try_into().unwrap())) };
+        // Start by ensuring the stack is large enough for the callee
+        for _ in caller_keyword_len..callee_kw_count {
+            argc += 1;
+            asm.stack_push(Type::Unknown);
+        }
+        // Now this is the stack_opnd() index to the 0th keyword argument.
+        let kwargs_stack_base = kwargs_order.len() as i32 - 1;
+
+        // Build the keyword rest parameter hash before we make any changes to the order of
+        // the supplied keyword arguments
+        if has_kwrest {
```

Comment:
Do you think it makes sense to carve out the kwrest logic as a separate function? It's generally hard to follow the sequence of operations in `gen_send_iseq`, so I wonder if it makes it easier to understand what happens after what in that function.

Reply:
Make sense. I think we should move all the keywords handling out, though. I'll make a PR.